File: container-with-most-water/njngwn.java
// Time Complexity: O(n), n: height.length
// Space Complexity: O(1)
class Solution {
    public int maxArea(int[] height) {
        int maxWaterAmount = 0;
        int leftLineIdx = 0;
        int rightLineIdx = height.length-1;

        while (leftLineIdx < rightLineIdx) {
            int leftHeight = height[leftLineIdx];
            int rightHeight = height[rightLineIdx];
            int tempAmount = 0;

            if (leftHeight < rightHeight) {
                tempAmount = leftHeight * (rightLineIdx - leftLineIdx);
                leftLineIdx++;
            } else {
                tempAmount = rightHeight * (rightLineIdx - leftLineIdx);
                rightLineIdx--;
            }

            // update maximum amount
            maxWaterAmount = tempAmount > maxWaterAmount ? tempAmount : maxWaterAmount;
        }

        return maxWaterAmount;
    }
}

File: container-with-most-water/njngwn.py
class Solution:
    # Time Complexity: O(n), n: len(height)
    # Space Complexity: O(1)
    def maxArea(self, height: List[int]) -> int:
        left, right = 0, len(height) - 1
        max_area = 0

        while left < right:
            w = right - left
            h = min(height[left], height[right])
            max_area = max(max_area, w * h)

            if height[left] <= height[right]:
                left += 1
            else:
                right -= 1

        return max_area
